cpp/.clang-format

```diff
@@ -0,0 +1,137 @@
+---
+# Refer to the following link for the explanation of each params:
+#   http://releases.llvm.org/8.0.0/tools/clang/docs/ClangFormatStyleOptions.html
+Language:        Cpp
+# BasedOnStyle:  Google
```

Comment:
Did you make any meaningful changes to the google C++ style guidelines? At a quick spot check, these seem to roughly match those. (Which seems great to me - nice to use an existing standard.) Would be great to comment on any deviations if they happen though.

Reply:
It roughly matches google style + some of the nice formattings I noticed during my time with xgboost.

Reply:
Have also updated this file with explanations to the ones I think are deviations. Let me know if I missed anything. This, now mostly matches google style.

Reply:
Resolving this one for now. Please file another comment, in case you find any issues.